### DISCUSSION TITLE
HTTP Error 422

### DISCUSSION THREAD

**Discussion by luciozambon** [fattureincloud/api#49] [Q&A] [answered]
Ricevo questo messaggio di errore: HTTP/1.1 422 
{"error":{"message":"Invalid request.","validation_result":{"data":["The i dati field is required."],"data.entity":["The data.entity field is required."]}}}

Ecco tutti i dettagli della richiesta (in cui ho oscurato i dettagli personali)
Psr7\Message::toString($e->getRequest()):

POST /c/[...]/issued_documents HTTP/1.1
User-Agent: GuzzleHttp/7
Host: api-v2.fattureincloud.it
Accept: application/json
Authorization: Bearer [secret token]

{
"data": {
    "type": "proforma",
    "subject": "",
    "visible_subject": "proforma test",
    "amount_net": 68.18,
    "amount_vat": 6.82,
    "amount_due_discount": 0,
    "entity": {
      "name": "Mary Red S.r.L.",
      "vat_number": "IT05432181211",
      "tax_code": "IT05432181211",
      "address_street": "Via Italia, 66",
      "address_postal_code": "20900",
      "address_city": "Milano",
      "address_province": "MI",
      "address_extra": "",
      "country": "Italia",
      "certified_email": "mary@pec.red.com",
      "ei_code": "ABCXCR1"
    },
    "date": "2022-08-10",
    "next_due_date": "2022-08-31",
    "items_list": [
      {
        "product_id": 333,
        "code": "SG3",
        "name": "Soggiorno",
        "measure": "",
        "net_price": 68.18182,
        "category": "",
        "gross_price": 75,
        "apply_withholding_taxes": true,
        "discount": 0,
        "discount_highlight": false,
        "in_dn": false,
        "qty": 1,
        "vat": {
          "id": 0,
          "value": 22,
          "description": ""
        },
        "stock": true,
        "description": "",
        "not_taxable": false
      }
    ],
    "payments_list": [
      {
        "amount": 75,
        "due_date": "2020-08-20",
         "payment_terms": {
          "days": 0,
          "type": "standard"
        },
        "status": "not_paid"
      }
    ],
  }
}

Psr7\Message::toString($e->getResponse()):

HTTP/1.1 422 Unprocessable Entity
Date: Thu, 11 Aug 2022 11:36:28 GMT
Content-Type: application/json
Content-Length: 156
Connection: keep-alive
Set-Cookie: AWSALB=[...]
Set-Cookie: AWSALBCORS=[...]
Server: Microsoft-IIS/8.5
Strict-Transport-Security: max-age=4838400; includeSubdomains;
Vary: Authorization
Cache-Control: no-cache, private
RateLimit-HourlyRemaining: 997
RateLimit-HourlyLimit: 1000
RateLimit-MonthlyRemaining: 19906
RateLimit-MonthlyLimit: 20000
X-Robots-Tag: noindex,nofollow,noarchive
X-AspNet-Version: 4.0.30319
X-Powered-By: ASP.NET

{"error":{"message":"Invalid request.","validation_result":{"data":["The i dati field is required."],"data.entity":["The data.entity field is required."]}}}

**Comment by valmoz** [comment, marked as answer]
Buongiorno,
inserendo il suo JSON d'esempio all'interno di un [JSON VALIDATOR](https://jsonformatter.curiousconcept.com/) è possibile notare che il JSON non è valido (c'è una virgola di troppo):
<img width="964" alt="image" src="https://user-images.githubusercontent.com/7254293/184133676-c68f8b35-52d4-4b81-b002-9aa8f3376a3a.png">
In generale, quando si ottiene l'errore "The i dati field is required" significa che c'è un errore nel decoding del JSON.
Lo so, pessimo messaggio di errore... ci stiamo lavorando.

Se possibile per la sua applicazione, le consiglio di utilizzare una delle nostre [SDK](https://developers.fattureincloud.it/docs/sdks/), che permettono di eliminare la maggior parte degli errori di formato come questo.

Grazie 
Mauro
